ambry-tools/src/main/java/com.github.ambry/tools/admin/SafeServerShutdownTool.java

```diff
@@ -0,0 +1,235 @@
+   * @throws InterruptedException if there are any interruptions while waiting for catch up.
+   * @throws IOException if there is any I/O error.
+   * @throws TimeoutException if there are timeout errors while making network requests.
+   */
+  public boolean prepareServerForShutdown(DataNodeId dataNodeId, long logGrowthPauseLagThresholdBytes, long timeoutSecs)
+      throws InterruptedException, IOException, TimeoutException {
+    long startTimeSecs = time.seconds();
+    boolean success = false;
+    // 1. Make sure the peers of the node have caught up till the logGrowthPauseLagThresholdBytes
+    LOGGER.info("Waiting until peers of {} are within {} bytes for all partitions", dataNodeId,
+        logGrowthPauseLagThresholdBytes);
+    if (waitTillCatchupOrTimeout(dataNodeId, logGrowthPauseLagThresholdBytes, startTimeSecs + timeoutSecs)) {
+      // 2. Disable all requests that result in log growth (PUT, DELETE and inbound replication).
+      LOGGER.info("Disabling PUT, DELETE and inbound replication on {} for all partitions", dataNodeId);
+      if (controlLogGrowth(dataNodeId, false)) {
+        // 3. Make sure the peers of the node have completely caught up
+        LOGGER.info("Waiting until peers of {} have completely caught up for all partitions", dataNodeId);
+        success = waitTillCatchupOrTimeout(dataNodeId, 0, startTimeSecs + timeoutSecs);
+      }
+    }
+    LOGGER.info("Server safe for shutdown: {}", success);
+    if (!success && !controlLogGrowth(dataNodeId, true)) {
+      throw new IllegalStateException("Could not reset state of " + dataNodeId + ". Please reset manually");
+    }
+    return success;
+  }
+
+  /**
+   * Waits until the lag of all peers of {@code dataNodeId} for all partitions <=
+   * {@code acceptableLagInBytes} or until time moves past {@code timeoutAtSecs}.
+   * @param dataNodeId the {@link DataNodeId} to ask for peer lag details.
+   * @param acceptableLagInBytes that lag in bytes that is considered "caught up".
+   * @param timeoutAtSecs the absolute time (in secs) at which the wait is considered failed.
+   * @return if the server's peers are confirmed to be within {@code acceptableLagInBytes}. {@code false} otherwise.
+   * @throws InterruptedException if there are any interruptions while waiting for catch up.
+   * @throws IOException if there is any I/O error.
+   * @throws TimeoutException if there are timeout errors while making network requests.
+   */
+  private boolean waitTillCatchupOrTimeout(DataNodeId dataNodeId, long acceptableLagInBytes, long timeoutAtSecs)
+      throws InterruptedException, IOException, TimeoutException {
+    boolean isCaughtUp;
+    do {
+      isCaughtUp = serverAdminTool.isCaughtUp(dataNodeId, null, acceptableLagInBytes).getSecond();
+      if (!isCaughtUp) {
+        // sleep for 5s
+        time.sleep(TimeUnit.SECONDS.toMillis(5));
+      }
+    } while (!isCaughtUp && time.seconds() < timeoutAtSecs);
```

Comment:
This looked like busy-waiting to me and was going to suggest something else, but saw you're basically sending a HTTP message and acting on the response each time so probably fine.  5 seconds seems kinda long though... you test to see how long it typically takes?  Does it usually go on the first try?

Reply:
It's hard to evaluate how long things typically take since that depends on several factors including size of cluster, number of partitions shared, speed of nodes that are replicating from you etc. But, as suggested, I have made this a config so it can be set based on empirical observations. 5s is a sensible default IMO.